hazelcast/src/test/java/com/hazelcast/spi/impl/operationexecutor/impl/OperationQueueImplTest.java

```diff
@@ -225,4 +228,71 @@ public void assertContent(Queue<Object> q, Object... expected) {
         List<Object> actual = new LinkedList<Object>(q);
         assertEquals(Arrays.asList(expected), actual);
     }
+
+    @Test
+    public void batchInterleaving_singleBatch() throws InterruptedException {
+        TaskBatch batch1 = createBatch();
+        Object operation = "operation";
+
+        // we add the batch and expect it to be taken first
+        operationQueue.add(batch1, false);
+        assertSame(batch1, operationQueue.take(false));
+
+        // while processing the batch, we add an operation
+        operationQueue.add(operation, false);
+
+        operationQueue.returnBatch(batch1);
+
+        // now we expect to pull the operation first before taking the returned batch
+        assertSame(operation, operationQueue.take(false));
+        assertSame(batch1, operationQueue.take(false));
+    }
+
+    private TaskBatch createBatch() {
+        return new TaskBatch(mock(PartitionTaskFactory.class), new BitSet(1), 1, 1);
+    }
+
+    @Test
+    public void batchInterleaving_singleBatch_andPriority() throws InterruptedException {
+        TaskBatch batch1 = createBatch();
+        Object operation = "operation";
+
+        // we add the batch and expect it to be taken first
+        operationQueue.add(batch1, false);
+        assertSame(batch1, operationQueue.take(false));
+
+        // while processing the batch, we add an operation
+        operationQueue.add(operation, true);
+
+        operationQueue.returnBatch(batch1);
+
+        // now we expect to pull the operation first before taking the returned batch
+        assertSame(operation, operationQueue.take(false));
+        assertSame(batch1, operationQueue.take(false));
+    }
+
+    @Test
+    public void batchInterleaving_multipleBatches() throws InterruptedException {
+        TaskBatch batch1 = createBatch();
+        TaskBatch batch2 = createBatch();
+        Object operation = "operation";
+
+        // we add the batch and expect it to be taken first
+        operationQueue.add(batch1, false);
+        operationQueue.add(batch2, false);
+        assertSame(batch1, operationQueue.take(false));
+
+        // while processing the batch, we add an operation
+        operationQueue.add(operation, false);
+
+        operationQueue.returnBatch(batch1);
+
+        // now we expect to pull the operation first before adding a ne batch
```

Comment:
typo: ne -> new; also this comment is quite confusing, it seems as copy-paste, I suggest to rewrite "pull the operation" to "pull the batch operation", WDYT?